pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go

```diff
@@ -130,10 +107,34 @@ func (c *FakeClient) CheckCloudEventsUnordered(t *testing.T, testName string, wa
 			}
 		}
 		if !found {
-			t.Errorf("unexpected event received: %q", event)
+			t.Errorf("unexpected event received: %q", got)
+		}
+	}
+}
+
+// readEventsFromChannel returns a slice of all events sent on the channel.
+// It fails the test if the incorrect number of events were received.
+func readEventsFromChannel(t *testing.T, c chan string, numExpectedEvents int) []string {
+	t.Helper()
+	var gotEvents []string
+	timeout := time.After(wait.ForeverTestTimeout)
+	// Read len(wantEvents) events from the channel, blocking until they have all been received
+	// or the test reaches its (long) timeout
+	for i := 0; i < numExpectedEvents; i++ {
+		select {
+		case got := <-c:
+			gotEvents = append(gotEvents, got)
+		case <-timeout:
+			t.Errorf("Expected %d events but timed out after receiving %d events", numExpectedEvents, len(gotEvents))
 		}
 	}
-	if len(expected) != 0 {
-		t.Errorf("%d events %#v are not received", len(expected), expected)
+	// Ensure that no extra events have been sent on the channel
+	for {
+		select {
+		case a := <-c:
+			t.Errorf("Unexpected event: %q", a)
+		default:
+			return gotEvents
+		}
```

Comment:
This will return immediately unless there is an extra event in the channel.

I'm not sure why in the first loop we block until timeout, but here we can return right away.
Unless we introduce some kind of "done writing" channel/message, shouldn't we would always have to wait for a timeout (which would not be ideal)?

Reply:
That said, we have a limited size buffer, so until we start reading messages from the channel, it won't be possible to write extra unexpected messages to the channel. Do you know if the channel will block then in case of overflow?

Reply:
Ah I think you're right, nice catch. I tried running a few test cases with extra events and they passed 😱 It looks like the referenced example is not quite right! FYI @Yongxuanzhang

Reply:
I can revert the concurrency changes for now and just have the cleanup-only changes in this PR if that's helpful?

Reply:
> Ah I think you're right, nice catch. I tried running a few test cases with extra events and they passed 😱 It looks like the referenced example is not quite right! FYI @Yongxuanzhang we may need to fix the k8s events tests to not pass when they shouldn't

For k8s event, this may not be an issue. Since we don't initialize the channel with size=expected events, it is set to 1000.

Reply:
IIUC, @afrittoli helped to find out this issue as well in https://github.com/tektoncd/pipeline/commit/fff3fcda2774f07d6787bbd0496c58610d381f9d, 

The comment is [here](https://github.com/tektoncd/pipeline/pull/5690#discussion_r1019626059). To address it I added a check when sending cloud event in FakeClient and will return error if extra events are to be sent

Reply:
I also tried increasing the channel size, and the tests still passed when they shouldn't. I think Andrea is right that the "default" branch gets picked up here since there may not be an event immediately ready on the channel.

Reply:
I think the check you added in response to Andrea's comment works because we waited until all events had been sent before reading any events from the channel, whereas this pattern starts reading from the channel without waiting for all the events to be available. I'm going to try to bring back some of the wait group logic in this PR but I'm not sure our logic is correct for k8s events.

Reply:
> I think the check you added in response to Andrea's comment works because we waited until all events had been sent before reading any events from the channel, whereas this pattern starts reading from the channel without waiting for all the events to be available. I'm going to try to bring back some of the wait group logic in this PR but I'm not sure our logic is correct for k8s events.

emmm let me take a look

Reply:
Oh I think there is similar situation in k8s event, but it should be very rare to happen. 

For k8s event, the channel size is set to 1000, so for our tests it's of low possibility that we can reach the buffer size and block the extra events. And since we have a large timer(equals to timeout of the test), we should be able to read the events even if it is sent in a goroutine and not in channel when we run `CheckEventsOrdered`.
 
I think maybe we can do the same for cloud events? Instead of setting the buffersize=expected events, we set it to a larger value (it was 100 before my previous PR)